Gems/Atom/Tools/ShaderManagementConsole/Code/Source/ShaderManagementConsoleApplication.cpp

```diff
@@ -93,14 +100,52 @@ namespace ShaderManagementConsole
 
         // Overriding default document type info to provide a custom view
         auto documentTypeInfo = ShaderManagementConsoleDocument::BuildDocumentTypeInfo();
-        documentTypeInfo.m_documentViewFactoryCallback = [this](const AZ::Crc32& toolId, const AZ::Uuid& documentId) {
-            return m_window->AddDocumentTab(documentId, new ShaderManagementConsoleTableView(toolId, documentId, m_window.get()));
+        documentTypeInfo.m_documentViewFactoryCallback = [this](const AZ::Crc32& toolId, const AZ::Uuid& documentId)
+        {
+            // Generic widget here serves to adapt the expected pointer type that AddDocumenTab takes.
+            // ShaderManagementConsoleContainer derives from Layout* so it wouldn't be compatible without using this dummy intermediary.
+            auto* container = new QWidget;
+            new ShaderManagementConsoleContainer(container, toolId, documentId, m_window.get());
+            return m_window->AddDocumentTab(documentId, container);
         };
         AtomToolsFramework::AtomToolsDocumentSystemRequestBus::Event(
             m_toolId, &AtomToolsFramework::AtomToolsDocumentSystemRequestBus::Handler::RegisterDocumentType, documentTypeInfo);
 
         m_window.reset(aznew ShaderManagementConsoleWindow(m_toolId));
         m_window->show();
+
+        using namespace AtomToolsFramework;
+        using namespace AzToolsFramework;
+        m_assetBrowserInteractions->RegisterContextMenuActions(
+            [](const AtomToolsAssetBrowserInteractions::AssetBrowserEntryVector& entries)
+            {
+                return entries.front()->GetEntryType() == AssetBrowser::AssetBrowserEntry::AssetEntryType::Source;
+            },
+            [this]([[maybe_unused]] QWidget* caller, QMenu* menu, const AtomToolsAssetBrowserInteractions::AssetBrowserEntryVector& entries)
+            {
+                const auto* entry = entries.empty() ? nullptr : entries.front();
+                if (!entry)
+                {
+                    return;
+                }
+                QFileInfo fileInfo(entry->GetFullPath().c_str());
+                QString extension = fileInfo.completeSuffix();
+                if (extension == "shader")
+                {
+                    AZStd::string savePath = entry->GetFullPath();
+                    AzFramework::StringFunc::Path::ReplaceExtension(savePath, AZ::RPI::ShaderVariantListSourceData::Extension);
+                    menu->addAction(
+                        "Create New Variant List (side by side to source)",
+                        [entry, savePath, this]()
+                        {
+                            AtomToolsFramework::AtomToolsDocumentSystemRequestBus::Event(
+                                m_toolId,
+                                &AtomToolsFramework::AtomToolsDocumentSystemRequestBus::Handler::CreateDocumentFromFilePath,
+                                entry->GetFullPath().c_str(),
+                                savePath);
```

Comment:
This looks like the open and save paths are the same. Should the save be empty or its extension changed to shadervariantlist? I could register a generic implementation of this menu command with the other asset browser document related commands that adds a “create X document” for files matching registered create from extensions.

Reply:
That behavior is already implemented. Since you registered “shader” as a create from extension, don't you already get the “create shader variant list document” by right clicking on a shader in the asset browser? 

\ShaderManagementConsoleDocument should be renamed to ShaderVariantListDocument. Whether or not it gets reimplemented, the idea was that a document class could potentially support opening multiple file types, like a text or image editor document type would support opening several text and image formats.

You can also register a generic document type for shader source data so that you can open it directly in the editor like material canvas. That would not interfere with you being able to right click and create a shader variant list from a shader source data document. They would work together.

    void MaterialCanvasApplication::InitShaderSourceDataDocumentType()
    {
        // Register document type for editing shader source data files. This document type does not have a central view
        // and will display a label widget that directs users to edit using the inspector.
        auto documentTypeInfo = AtomToolsFramework::AtomToolsAnyDocument::BuildDocumentTypeInfo(
            "Shader Source Data",
            { "shader" },
            {},
            AZStd::any(AZ::RPI::ShaderSourceData()),
            AZ::RPI::ShaderSourceData::TYPEINFO_Uuid()); // Supplying ID because it is not included in the JSON file

        documentTypeInfo.m_documentViewFactoryCallback = [this]([[maybe_unused]] const AZ::Crc32& toolId, const AZ::Uuid& documentId)
        {
            auto viewWidget = new QLabel("Shader Source Data properties can be edited in the inspector.", m_window.get());
            viewWidget->setAlignment(Qt::AlignCenter);
            return m_window->AddDocumentTab(documentId, viewWidget);
        };

        AtomToolsFramework::AtomToolsDocumentSystemRequestBus::Event(
            m_toolId, &AtomToolsFramework::AtomToolsDocumentSystemRequestBus::Handler::RegisterDocumentType, documentTypeInfo);
    }

Reply:
Yes we have a double behavior, the generic tool system already adds a "create new variant list" in the context menu of the asset browser, which is good. But it doesn't give a method to the client tool to specify the save path, thus it opens a file picker which risks to puzzle the user because it starts in %home%/document path so probably not where it needs to be.
If we can fix that, it would remove the need for the custom extension

Reply:
> This looks like the open and save paths are the same. Should the save be empty or its extension changed to shadervariantlist? I could register a generic implementation of this menu command with the other asset browser document related commands that adds a “create X document” for files matching registered create from extensions.

The `savePath` extension is changed line 136 to .shadervariantlist